ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpEngine.kt

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.engine.java
+
+import io.ktor.client.engine.*
+import io.ktor.client.features.*
+import io.ktor.client.request.*
+import io.ktor.client.utils.*
+import io.ktor.util.*
+import kotlinx.coroutines.*
+import java.io.*
+import java.net.*
+import java.net.http.*
+import java.time.*
+import java.util.concurrent.*
+import kotlin.coroutines.*
+
+@InternalAPI
+@Suppress("KDocMissingDocumentation")
+public class JavaHttpEngine(override val config: JavaHttpConfig) : HttpClientEngineBase("ktor-java") {
+
+    public override val dispatcher: CoroutineDispatcher by lazy {
+        Dispatchers.clientDispatcher(
+            config.threadsCount,
+            "ktor-java-dispatcher"
+        )
+    }
+
+    public override val supportedCapabilities: Set<HttpTimeout.Feature> = setOf(HttpTimeout)
+
+    override val coroutineContext: CoroutineContext
+
+    private val requestsJob = SilentSupervisor(super.coroutineContext[Job])
+
+    private var httpClient: HttpClient? = null
+
+    /**
+     * Exposed for tests only.
+     */
+    internal val executor = Executors.newCachedThreadPool {
+        Thread(it, "ktor-client-java").apply {
+            isDaemon = true
+            setUncaughtExceptionHandler { _, _ -> }
+        }
+    }
+
+    init {
+        coroutineContext = super.coroutineContext + requestsJob + CoroutineName("java-engine")
+
+        @OptIn(ExperimentalCoroutinesApi::class)
+        GlobalScope.launch(super.coroutineContext, start = CoroutineStart.ATOMIC) {
+            try {
+                requestsJob[Job]!!.join()
+            } finally {
+                httpClient = null
+                (dispatcher as Closeable).close()
+                executor.shutdown()
+            }
+        }
+    }
+
+    override suspend fun execute(data: HttpRequestData): HttpResponseData {
+        val engine = getJavaHttpClient(data)
+        val callContext = callContext()
+
+        return if (data.isUpgradeRequest()) {
+            engine.executeWebSocketRequest(callContext, data)
+        } else {
+            engine.executeHttpRequest(callContext, data)
+        }
+    }
+
+    override fun close() {
+        super.close()
+
+        (requestsJob[Job] as CompletableJob).complete()
+    }
+
+    private fun getJavaHttpClient(data: HttpRequestData): HttpClient {
+        return httpClient ?: synchronized(this) {
```

Comment:
I would extract double-check with synchronized in the separate method, but I don't insist

Reply:
I tried to find a more readable way to do that, but failed. Please feel free to fix it once it'll be merged.